### ISSUE TITLE
Add layer(s): Google Maps

### ISSUE THREAD

**Issue by mstenta** [farmOS/farmOS-map#8]
See https://github.com/mapgears/ol3-google-maps

**Comment by mstenta** [comment]
> There will need to be a way of passing a Google Maps API key to the library.

Correction: we may not need to do this after all. Instead, whoever is using the library will need to add the following script to the page before creating maps, which will include the key:

`<script type="text/javascript" src="https://maps.googleapis.com/maps/api/js?v=3&key=<KEY>"></script>`

So it will be up to the user of the farmOS-map library to add that, which means we don't need to manage the key within the library itself.

**Comment by mstenta** [comment]
I have a working branch started: https://github.com/mstenta/farmOS-map/tree/google

Still want to give some thought to how this work...

**Comment by mstenta** [comment]
Couple of notes:

* Currently Google Maps layers cannot be put inside a Group layer. See this issue in the `ol3-google-maps` queue: https://github.com/mapgears/ol3-google-maps/issues/37
* Right now all of the Google Maps layers are turned off by default. It would be nice to add some way of defining the default base layer for a map (and maybe a way for other JS code to change the default - ala via Drupal modules). This may necessitate a more structured way of managing the layers in a map, because right now they are just an array, and are not named, so it would be hard to programmatically change which layer is the default.
* There seem to be some style inconsistencies when you toggle back and forth between the Google layers and the OSM layer. Specifically:
    * OL control buttons and layer switcher panel are smaller when Google layers are visible.
    * Font is different in the layer switcher panel.

**Comment by mstenta** [comment]
Another issue I'm noticing: when a GeoJSON layer is added to the map, and you switch to a Google layer, it becomes impossible to disable the GeoJSON layer. Unchecking it in the layer switcher does not hide the GeoJSON feature. However, switching back to the OSM base layer reveals that the GeoJSON layer was indeed hidden. Switching back to a Google layer after that: the GeoJSON layer is hidden there as well. So it appears that turning GeoJSON layers on/off has no effect when a Google layer is visible.

**Comment by mstenta** [comment]
Opened a pull request for the basic integration: https://github.com/farmOS/farmOS-map/pull/33

**Comment by mstenta** [comment]
> when a GeoJSON layer is added to the map, and you switch to a Google layer, it becomes impossible to disable the GeoJSON layer

Came across this: https://github.com/mapgears/ol3-google-maps/pull/165#issuecomment-285179241

Sounds somewhat similar - perhaps worth testing the fix they contributed.

**Comment by mstenta** [comment]
I dusted this off and discovered ol3-google-maps isn't compatible with OpenLayers v6.0.0+ (which we have upgraded to in the meantime). I created a PR to fix this upstream: https://github.com/mapgears/ol3-google-maps/pull/283

Notably, this doesn't seem to be an issue anymore: 

> when a GeoJSON layer is added to the map, and you switch to a Google layer, it becomes impossible to disable the GeoJSON layer

Not sure if that was from the OL6 upgrade or what.  Kinda surprised it just went away like that... fingers crossed!

**Comment by mstenta** [comment]
Hmm seeing another issue now: the GeoJSON layers are not lining up with the Google layers, and they move around as you pan/zoom the map.

**Comment by mstenta** [comment]
> the GeoJSON layers are not lining up with the Google layers, and they move around as you pan/zoom the map

Ahh... I remember what causes this now: https://www.drupal.org/project/openlayers/issues/2946213

tl;dr: we need to set `constrainResolution: true` on the map view, so that only discreet zoom levels are used (not fractional levels).

I wonder if I can set it up so that *only* happens if you enable Google layers. Because otherwise fractional zoom levels are really nice.

**Comment by mstenta** [comment]
> we need to set constrainResolution: true on the map view

> set it up so that only happens if you enable Google layers

Done and done!  Working much better now.  :-)

**Comment by mstenta** [comment]
So I think there are just two remaining issues:

1. Google layers don't work inside a layer group: https://github.com/mapgears/ol3-google-maps/issues/285
2. Style inconsistencies when you toggle back and forth between the Google and non-Google layers (see comment above: https://github.com/farmOS/farmOS-map/issues/8#issuecomment-511567553).

**Comment by mstenta** [comment]
> Style inconsistencies when you toggle back and forth between the Google and non-Google layers

I traced this to the following inline CSS rule:

```css
.gm-style {
    font: 400 11px Roboto, Arial, sans-serif;
    text-decoration: none;
}
```

Found a relevant thread here: https://stackoverflow.com/questions/18281770/disable-css-styles-in-google-maps-3-14-infowindow

**Comment by mstenta** [comment]
OK another clue:

The inline CSS rule that Google Maps adds is *sometimes* added in the document HEAD, and *sometimes* it is added inside the map DIV itself. The `ol3-google-maps` library also adds its own inline styles in the document HEAD, which attempt to override the `font` style and change it to `inherit`. Depending on where the `gm-style` inline CSS is added (HEAD or DIV) determines whether or not the `ol3-google-maps` styles will actually override them properly.

So:
- If you load the map with a Google layer enabled, the `gm-style` CSS is added to the HEAD, which allows the `ol3-google-maps` CSS to override them, which results in **correct** icon font family and size.
- If you load the map with a NON-Google layer enabled, and then toggle to a Google layer, the `gm-style` CSS is added to the map DIV, which does NOT allow the `ol3-google-maps` CSS to override them, with results in **incorrect** icon font family and size.

So, if we could somehow ensure that the `gm-style` CSS were added to the document HEAD no matter what, that should fix this issue.

**Comment by mstenta** [comment]
> So, if we could somehow ensure that the `gm-style` CSS were added to the document HEAD no matter what, that should fix this issue.

Or, we could just throw our own `!important` style in there and brute force it. :-/

FWIW, there are bug reports for this in the Google issue tracker:

https://issuetracker.google.com/issues/35824012

https://issuetracker.google.com/issues/35823563

And this: https://groups.google.com/forum/#!topic/google-maps-js-api-v3/zBQ-IL5nuWs

Last updates were a loooong time ago, though (2014), so I don't know if we can expect any more movement on that.

**Comment by mstenta** [comment]
I think we may need these two ol3-google-maps upstream issues as well:

- Support Clusters https://github.com/mapgears/ol3-google-maps/issues/52
- Support for GeometryCollection in VectorFeatureHerald https://github.com/mapgears/ol3-google-maps/issues/255

**Comment by mstenta** [comment]
Just for future reference, I found this thread that shows how to include Google layers as a simple XYZ source: https://github.com/openlayers/openlayers/issues/9900

Important to note @ahocevar's comment, however:

> Needless to say that the above is a violation of Google's Terms of Use.

I am assuming that ol3-google-maps is not in violation of GToU because it uses the actual Google Maps JS library alongside OpenLayers JS.

**Comment by mstenta** [comment]
> Google layers don't work inside a layer group

I've debugged this upstream and got it working, for the most part: https://github.com/mapgears/ol3-google-maps/issues/285

There are still some remaining issues, but I'm not sure if they affect farmOS. We'll see...

I created a new `farmOS-map` branch on my fork of ol3-google-maps, which also merges in the OL6 pull request (https://github.com/mapgears/ol3-google-maps/issues/285): https://github.com/mstenta/ol3-google-maps/tree/farmOS-map

I plan to reference this branch in the farmOS-map `package.json` until those issues are resolved/merge upstream.

So I think the only remaining pieces are the ones mentioned in my comment above: https://github.com/farmOS/farmOS-map/issues/8#issuecomment-570789476

**Comment by mstenta** [comment]
Ah ha! I may have discovered a simple trick that will solve the cluster and geometrycollection issues: https://github.com/mapgears/ol3-google-maps/issues/52#issuecomment-570809053

Basically, we can tell ol3-google-maps NOT to manage vector layers - so they will be rendered by OpenLayers only, and displayed on top of the Google layers. Testing this more now to be sure...

**Comment by mstenta** [comment]
An old issue with Google layer zoom levels that I "fixed" a few years ago came up again, so I "fixed" it again. It's still not perfect, but it's no better/worse than it has been at least. More info: https://github.com/mapgears/ol3-google-maps/issues/216

**Comment by mstenta** [comment]
Discovered that Google Maps do not work with map rotation. Created a separate bug report: https://github.com/farmOS/farmOS-map/issues/59

Not sure if I'm going to try to fix that before release, or leave it as a "known issue". Rotation works fine if you are not using the Google layers, and it's a nice feature to have IMO (also was a requested feature: https://github.com/farmOS/farmOS-map/issues/39).

**Comment by mstenta** [comment]
Merged #33 and released v0.9.0: https://github.com/farmOS/farmOS-map/releases/tag/v0.9.0
